### PR TITLE
BUG: error calculating BusinessHourMixin.apply for long business hour per day

### PR DESCRIPTION
**Pull request by jiangyue12392**
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew: handle the case where adding of business hour to the timestamp results in an ending time landing in the next business time period. Example

Input:
```python
bh = BusinessHour(n=4, start='00:00', end='23:00')
dt = datetime(2014, 7, 3, 22)
dt + bh
```
Expected:
```python
datetime(2014, 7, 4, 3)
```

### CODE REVIEW

**Comment by gfyoung** [comment]
@jiangyue12392 : Thanks for submitting this!  For those tests cases that you've added, what happens currently without your fixes?

**Comment by jiangyue12392** [comment]
@gfyoung The current code without fixes would return a wrong value if the adding of business hour to the timestamp results in an ending time landing in the next business time period. For example, if the working hour is 00:00 to 23:00 and 4 business hours are added at 22:00, the current code would return 02:00 in the next day instead of 03:00 the next business day.

**Comment by gfyoung** [comment]
@jiangyue12392 : I see.  For clarify, could you show the output of those examples, using a format like this (we use this format for our issue tracker):

# Example
~~~python
print("Hello World")
~~~

This outputs:

~~~
Hello World
~~~

**Comment by jiangyue12392** [comment]
@gfyoung Example inputs and outputs:
Input:
```python
bh = BusinessHour(n=4, start='00:00', end='23:00')
dt = datetime(2014, 7, 3, 22)
dt + bh
```
Expected:
```python
datetime(2014, 7, 4, 3)
```
Current Output:
```python
datetime(2014, 7, 4, 2)
```

**Comment by gfyoung** [comment]
Got it.  Can you add a `whatsnew` entry as well?

**Comment by jiangyue12392** [comment]
@gfyoung `whatsnew` entry added

**Comment by gfyoung** [comment]
@jiangyue12392 : A `whatsnew` entry is adding a bullet point to this file:

https://github.com/pandas-dev/pandas/blob/master/doc/source/whatsnew/v0.25.0.rst

I recommend adding it under the `Timedelta` section.

**Comment by jiangyue12392** [comment]
@gfyoung new entry added

File: pandas/tests/tseries/offsets/test_offsets.py [hunk 1]
```diff
@@ -1287,6 +1287,23 @@ def test_opening_time(self, case):
         datetime(2014, 7, 7, 19, 30): datetime(2014, 7, 5, 4, 30),
         datetime(2014, 7, 7, 19, 30, 30): datetime(2014, 7, 5, 4, 30, 30)}))
 
+    # long business hours
```
Comment (gfyoung) [inline pandas/tests/tseries/offsets/test_offsets.py]: Reference this PR as part of this comment e.g. add:

`see gh-26381`

To the end of that comment.

**Review by gfyoung — APPROVED** [review]
cc @jreback @mroeschke

File: pandas/tests/tseries/offsets/test_offsets.py (hunk 1 above)
Comment (jiangyue12392) [reply pandas/tests/tseries/offsets/test_offsets.py]: Reference done

File: doc/source/whatsnew/v0.25.0.rst [hunk 2]
```diff
@@ -285,7 +285,7 @@ Timedelta
 
 - Bug in :func:`TimedeltaIndex.intersection` where for non-monotonic indices in some cases an empty ``Index`` was returned when in fact an intersection existed (:issue:`25913`)
 - Bug with comparisons between :class:`Timedelta` and ``NaT`` raising ``TypeError`` (:issue:`26039`)
--
+- Bug in adding of business hour to the timestamp with an ending time landing in the next business time period (:pull:`26381`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v0.25.0.rst]: ``Bug when adding or subtracting a :class:`BusinessHour` to a :class:`Timestamp` with the resulting time landing in the next or prior day respectively (:issue:`26381`)``

**Comment by jreback** [comment]
looks fine, ex- @mroeschke comment on the whatsnew.

File: pandas/tests/tseries/offsets/test_offsets.py [hunk 3]
```diff
@@ -1287,6 +1287,23 @@ def test_opening_time(self, case):
         datetime(2014, 7, 7, 19, 30): datetime(2014, 7, 5, 4, 30),
         datetime(2014, 7, 7, 19, 30, 30): datetime(2014, 7, 5, 4, 30, 30)}))
 
+    # long business hours (see gh-26381)
+    apply_cases.append((BusinessHour(n=4, start='00:00', end='23:00'), {
```
Comment (jreback) [inline pandas/tests/tseries/offsets/test_offsets.py]: would love a separate PR to split up test_offsets.py a bit :->

**Comment by jreback** [comment]
lgtm. ping on green.

**Comment by jiangyue12392** [comment]
@jreback

**Comment by TomAugspurger** [comment]
Thanks @jiangyue12392, nice work!

**Pull request merged**